src/Core/InstanceConfigTrait.php

```diff
@@ -110,11 +120,11 @@ public function setConfig($key, $value = null, $merge = true)
      * $this->getConfig('some-key', 'default-value');
      * ```
      *
-     * @param string|null $key The key to get or null for the whole config.
-     * @param mixed $default The return value when the key does not exist.
-     * @return mixed Configuration data at the named key or null if the key does not exist.
+     * @param string $key The key to get.
+     * @param mixed|null $default The return value when the key does not exist.
+     * @return mixed|null Configuration data at the named key or null if the key does not exist.
      */
-    public function getConfig(?string $key = null, $default = null)
+    public function getConfig(string $key, $default = null)
```

Comment:
This is a pretty notable breaking change that we haven't had a deprecation warning for, which is rough for users.

Reply:
We can backport some of it into 3.x.
I can see the impact on user land code. We have to decide if it is worth it, given the benefits of the typehinting.

Reply:
I generally agree that typehinting improves the code. This adds a single typehint, but at the cost of breaking a pretty core method.

Reply:
Pretty is relative, the current way allows dangerous bugs like above, silently. :)

Reply:
I too don't like this change. Splitting a method which already takes just a single type + null as argument is going too far.

Back porting

Reply:
It is the only way to catch the silent bug of reading all config by accident as above.

Reply:
In the example you have provided the argument to `getConfig()` is a static string, so how does preventing `null` argument relate to that?  In my experience it's quite rare to pass a variable to `getConfig()`, which could unintentionally have null value.

Also you have added the new `getConfigOrFail()` which doesn't allow null argument so one can just use that.

Reply:
Those are two different things.
I am OK with this going into a 5.0 cleanup though, if we want to postpone.
But IMO with rector this should also be solvable at this point.